File: src/python/flashloanarbitrage.py
import os
import time
from dotenv import load_dotenv
from web3 import Web3

# Load environment variables
load_dotenv()
SEPOLIA_URL = os.getenv("Sepolia_Url")
PRIVATE_KEY = os.getenv("Private_Key")
WALLET_ADDRESS = os.getenv("Wallet")
CONTRACT_ADDRESS = os.getenv("Flashloan_Arbitrage")
DEX_ADDRESS = os.getenv("Dex")
USDC_ADDRESS = os.getenv("Sepolia_USDC")
DAI_ADDRESS = os.getenv("Sepolia_DAI")

# Connect to Sepolia network
web3 = Web3(Web3.HTTPProvider(SEPOLIA_URL))
if not web3.is_connected():
    print("Failed to connect to Sepolia network")
    exit()

# FlashLoanArbitrage ABI (partial for flash loan and approve functions)
FLASHLOAN_ABI = [
    {
        "constant": False,
        "inputs": [
            {"name": "_token", "type": "address"},
            {"name": "_amount", "type": "uint256"},
            {"name": "dexAddress", "type": "address"},
            {"name": "tokenAAddress", "type": "address"},
            {"name": "tokenBAddress", "type": "address"},
            {"name": "depositAmount", "type": "uint256"}
        ],
        "name": "requestFlashLoan",
        "outputs": [],
        "type": "function"
    }
]

# Initialize contract instance
flashloan_contract = web3.eth.contract(address=CONTRACT_ADDRESS, abi=FLASHLOAN_ABI)

def request_flash_loan(token_address, amount, dex_address, token_a, token_b, deposit_amount):
    try:
        print("Preparing transaction to request flash loan...")

        # Encode transaction data
        transaction_data = flashloan_contract.functions.requestFlashLoan(
            token_address,
            amount,
            dex_address,
            token_a,
            token_b,
            deposit_amount
        ).build_transaction({
            'from': WALLET_ADDRESS
        })

        # Log the transaction data
        print(f"Transaction Data: {transaction_data['data']}")

        # Fetch gas price and nonce
        gas_price = web3.eth.gas_price
        nonce = web3.eth.get_transaction_count(WALLET_ADDRESS, 'pending')

        # Estimate gas dynamically
        gas_estimate = flashloan_contract.functions.requestFlashLoan(
            token_address, amount, dex_address, token_a, token_b, deposit_amount
        ).estimate_gas({'from': WALLET_ADDRESS})

        print(f"Gas Price: {gas_price}")
        print(f"Gas Estimate: {gas_estimate}")
        print(f"Nonce: {nonce}")

        # Build transaction
        tx = {
            'from': WALLET_ADDRESS,
            'to': CONTRACT_ADDRESS,
            'data': transaction_data['data'],
            'gas': gas_estimate,
            'gasPrice': gas_price,
            'nonce': nonce
        }

        print("Signing the transaction...")

        # Sign the transaction
        signed_tx = web3.eth.account.sign_transaction(tx, PRIVATE_KEY)

        print("Sending the transaction...")

        # Send transaction
        tx_hash = web3.eth.send_raw_transaction(signed_tx.rawTransaction)

        print(f"Transaction sent! Hash: {web3.to_hex(tx_hash)}")

        # Wait for receipt with a longer timeout
        print("Waiting for transaction receipt...")
        tx_receipt = web3.eth.wait_for_transaction_receipt(tx_hash, timeout=600)  # 10-minute timeout

        if tx_receipt.status == 1:
            print("Flash loan arbitrage transaction successful!")
        else:
            print("Flash loan arbitrage transaction failed!")

    except Exception as e:
        print("Error during flash loan request:")
        print(f"Error Type: {type(e).__name__}")
        print(f"Error Message: {e}")

# Main function
if __name__ == "__main__":
    # Flash loan parameters
    flash_loan_amount = web3.to_wei("1000", "mwei")  # 1000 USDC (6 decimals)
    deposit_amount = flash_loan_amount  # Same as flash loan amount

    request_flash_loan(
        token_address=USDC_ADDRESS,
        amount=flash_loan_amount,
        dex_address=DEX_ADDRESS,
        token_a=USDC_ADDRESS,
        token_b=DAI_ADDRESS,
        deposit_amount=deposit_amount
    )

File: src/javascript/flashloanarbitrage.js
require("dotenv").config();
const { Web3 } = require("web3");

// Load environment variables
const PRIVATE_KEY = process.env.Private_Key;
const WALLET_ADDRESS = process.env.Wallet;
const FLASHLOAN_CONTRACT_ADDRESS = process.env.Flashloan_Arbitrage;
const SEPOLIA_RPC_URL = process.env.Sepolia_Url;
const DEX_ADDRESS = process.env.Dex;
const USDC_ADDRESS = process.env.Sepolia_USDC;
const DAI_ADDRESS = process.env.Sepolia_DAI;

// Initialize Web3
const web3 = new Web3(new Web3.providers.HttpProvider(SEPOLIA_RPC_URL));

// FlashLoanArbitrage Contract ABI
const FLASHLOAN_ABI = [
  {
    constant: false,
    inputs: [
      { name: "_token", type: "address" },
      { name: "_amount", type: "uint256" },
      { name: "dexAddress", type: "address" },
      { name: "tokenAAddress", type: "address" },
      { name: "tokenBAddress", type: "address" },
      { name: "depositAmount", type: "uint256" },
    ],
    name: "requestFlashLoan",
    outputs: [],
    type: "function",
  },
];

// FlashLoanArbitrage Contract Instance
const flashloanContract = new web3.eth.Contract(
  FLASHLOAN_ABI,
  FLASHLOAN_CONTRACT_ADDRESS
);

// Function to send a transaction with timeout
const sendTransactionWithTimeout = (rawTransaction, timeout) => {
  return Promise.race([
    web3.eth.sendSignedTransaction(rawTransaction),
    new Promise((_, reject) =>
      setTimeout(() => reject(new Error("Transaction timeout")), timeout)
    ),
  ]);
};

// Main function to request a flash loan
async function requestFlashLoan() {
  try {
    console.log("Preparing transaction to request flash loan...");

    // Define parameters
    const tokenA = USDC_ADDRESS; // USDC
    const tokenB = DAI_ADDRESS; // DAI
    const dexAddress = DEX_ADDRESS; // DEX contract address
    const flashLoanAmount = web3.utils.toWei("1000", "mwei"); // 1000 USDC (6 decimals)
    const depositAmount = flashLoanAmount; // Same as flash loan amount

    console.log(`Token A (USDC): ${tokenA}`);
    console.log(`Token B (DAI): ${tokenB}`);
    console.log(`DEX Address: ${dexAddress}`);
    console.log(`Flash Loan Amount: ${flashLoanAmount}`);
    console.log(`Deposit Amount: ${depositAmount}`);

    // Build transaction data
    const data = flashloanContract.methods
      .requestFlashLoan(tokenA, flashLoanAmount, dexAddress, tokenA, tokenB, depositAmount)
      .encodeABI();

    // Fetch gas price, gas estimate, and nonce
    const gasPrice = await web3.eth.getGasPrice();
    const gasEstimate = await web3.eth.estimateGas({
      from: WALLET_ADDRESS,
      to: FLASHLOAN_CONTRACT_ADDRESS,
      data: data,
    });
    const nonce = await web3.eth.getTransactionCount(WALLET_ADDRESS, "pending");

    console.log(`Gas Price: ${gasPrice}`);
    console.log(`Gas Estimate: ${gasEstimate}`);
    console.log(`Nonce: ${nonce}`);

    // Build transaction
    const tx = {
      from: WALLET_ADDRESS,
      to: FLASHLOAN_CONTRACT_ADDRESS,
      data: data,
      gas: gasEstimate,
      gasPrice: gasPrice,
      nonce: nonce,
    };

    // Sign the transaction
    console.log("Signing the transaction...");
    const signedTx = await web3.eth.accounts.signTransaction(tx, PRIVATE_KEY);

    console.log("Sending the transaction...");

    // Send the transaction with a timeout
    const receipt = await sendTransactionWithTimeout(signedTx.rawTransaction, 30000); // 30s timeout

    console.log("Flash loan request successful!");
    console.log(`Transaction Hash: ${receipt.transactionHash}`);
    console.log("Transaction Receipt:", receipt);
  } catch (error) {
    console.error("Error while sending the transaction:", error);
  }
}

// Execute the function
requestFlashLoan();
